mapper/orm/src/main/java/org/hibernate/search/mapper/orm/outbox/impl/OutboxEventBackgroundExecutor.java

```diff
@@ -0,0 +1,283 @@
+import org.hibernate.search.engine.reporting.EntityIndexingFailureContext;
+import org.hibernate.search.engine.reporting.FailureHandler;
+import org.hibernate.search.mapper.orm.automaticindexing.spi.AutomaticIndexingMappingContext;
+import org.hibernate.search.mapper.orm.logging.impl.Log;
+import org.hibernate.search.mapper.pojo.route.DocumentRouteDescriptor;
+import org.hibernate.search.mapper.pojo.route.DocumentRoutesDescriptor;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+import org.hibernate.search.util.common.serialization.spi.SerializationUtils;
+
+public class OutboxEventBackgroundExecutor {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+	private static final int MAX_RETRIES = 3;
+
+	private enum Status {
+		STOPPED,
+		STARTED
+	}
+
+	private final AutomaticIndexingMappingContext mapping;
+	private final int pollingInterval;
+	private final int batchSize;
+	private final AtomicReference<Status> status = new AtomicReference<>( Status.STOPPED );
+	private final FailureHandler failureHandler;
+	private final SingletonTask processingTask;
+
+	public OutboxEventBackgroundExecutor(AutomaticIndexingMappingContext mapping, ScheduledExecutorService executor,
+			int pollingInterval, int batchSize) {
+		this.mapping = mapping;
+		this.pollingInterval = pollingInterval;
+		this.batchSize = batchSize;
+
+		failureHandler = mapping.failureHandler();
+		processingTask = new SingletonTask(
+				"Delayed commit for " + OutboxTableAutomaticIndexingStrategy.NAME,
+				new HibernateOrmOutboxWorker(),
+				new HibernateOrmOutboxScheduler( executor ),
+				failureHandler
+		);
+	}
+
+	public void start() {
+		status.set( Status.STARTED );
+		processingTask.ensureScheduled();
+	}
+
+	public CompletableFuture<?> completion() {
+		status.set( Status.STOPPED );
+		return processingTask.completion();
+	}
+
+	public void stop() {
+		processingTask.stop();
+	}
+
+	private class HibernateOrmOutboxWorker implements SingletonTask.Worker {
+
+		@Override
+		public CompletableFuture<?> work() {
+			if ( mapping.sessionFactory().isClosed() ) {
+				log.sessionFactoryIsClosedOnOutboxProcessing();
+				return CompletableFuture.completedFuture( null );
+			}
+
+			try ( Session session = mapping.sessionFactory().openSession() ) {
+				withinTransaction( session, () -> {
+					// TODO HSEARCH-4194 Guarantee a minimum delay on processing retries of outbox events
+					List<OutboxEvent> outboxRetries = findOutboxRetries( session );
+					if ( !outboxRetries.isEmpty() ) {
+						// Process the event retries
+						OutboxEventProcessingPlan eventProcessing = new OutboxEventProcessingPlan(
+								mapping, session, outboxRetries );
+						eventProcessing.processEvents();
+						deleteProcessedOutboxRetries( session, outboxRetries, eventProcessing.collectFailedEvents() );
+						updateUnprocessedOutboxRetries( session, eventProcessing );
+					}
+				} );
```

Comment:
With just a few changes, you can remove this block of code along with a few methods, then use the same code for retries and initial events, and all the tests will still pass.

I pushed a commit to demonstrate that, please keep it.

These changes also ensure that a retry event always has an ID higher than the initial event, which could prove useful.